Add active_units configuration

cjwatson opened the pull request:
Assuming that the requiring charms support it, this allows reasonably
graceful migration from one NFS unit to another, along the lines of the
following procedure:

```
$ juju config nfs active_units=<old unit ID>
$ juju add-unit nfs
$ juju config nfs active_units=none
# wait for clients to unmount, then move storage to new unit
$ juju config nfs active_units=<new unit ID>
# wait for clients to mount
$ juju remove-unit <old unit ID>
$ juju config nfs active_units=
```

This involves moving publishing data on the mount relation to
`read_nfs_mounts`, which seems more correct anyway (we shouldn't publish
a mount until it has been exported).

cjwatson committed:
Add active_units configuration

Assuming that the requiring charms support it, this allows reasonably
graceful migration from one NFS unit to another, along the lines of the
following procedure:

  $ juju config nfs active_units=<old unit ID>
  $ juju add-unit nfs
  $ juju config nfs active_units=none
  # wait for clients to unmount, then move storage to new unit
  $ juju config nfs active_units=<new unit ID>
  # wait for clients to mount
  $ juju remove-unit <old unit ID>
  $ juju config nfs active_units=

This involves moving publishing data on the mount relation to
`read_nfs_mounts`, which seems more correct anyway (we shouldn't publish
a mount until it has been exported).

hyperbolic2346 commented:
Can we get a documentation update to talk about this config option and how one would use it? I think pretty much a slightly more detailed version of your initial comment would be fine.

cjwatson committed:
Add storage migration documentation

cjwatson commented:
Sure - I've added some documentation to README.md.  It might be best to wait until the discussion in https://github.com/juju-solutions/interface-mount/pull/5 is complete before merging this, though, as depending on how that goes it might necessitate a different approach here.

cjwatson committed:
Tweak how read_nfs_mounts gets the mount endpoint

The `endpoint.nfs.joined` flag isn't necessarily set when
`read_nfs_mounts` is called, so use endpoint_from_name('nfs') instead.

hyperbolic2346 commented:
For the future, endpoint_from_flag works via the same names you would use for a @when clause. Something like `endpoint_from_flag('nfs.available')` or the like.

cjwatson committed:
Publish active unit from all units

In order that the mount interface and requiring charms don't need to
change to support `active_units` migration, we now use a peer relation
to find the address of the first active unit and publish that from all
units.

cjwatson committed:
Note abruptness of migration

cjwatson commented:
I've reworked this somewhat in light of @johnsca's comments in https://github.com/juju-solutions/interface-mount/pull/5.  The usage pattern remains the same, but the way in which data is published on the relation changes.

https://github.com/juju-solutions/interface-mount/pull/5 is no longer required to make this work (although I think it would still be a reasonable cleanup), but https://github.com/juju-solutions/interface-mount/pull/6 is now required in order to be able to publish a `hostname`.

hyperbolic2346 commented:
This looks ok to me, but since it is in response to @johnsca's comments, I would like his opinion here.

johnsca commented on reactive/nfs.py:
```diff
@@ -50,27 +54,96 @@ def update_config():
+          'config.changed.active_units')
 def read_nfs_mounts():
+    mount_interface = endpoint_from_name('nfs')
+    if mount_interface is None:
```
This check could be replaced with a `@when('endpoint.nfs.joined')` decorator on the handler.

johnsca commented on reactive/nfs.py:
```diff
@@ -50,27 +54,96 @@ def update_config():
+    storage_root = config.get('storage_root')
+    mount_options = config.get('mount_options')
+    active_ip = None
+    active_units = config.get('active_units', '')
```
Since there are default values provided in `config.yaml`, technically this (and the others) could just be `config['active_units']`, but that's mostly just style.  I would at least remove the `, ''` bit, since it might make someone down the road think a value could be placed there (but it would never be used).

johnsca commented on reactive/nfs.py:
```diff
@@ -50,27 +54,96 @@ def update_config():
+        peer_endpoint = endpoint_from_name('peer')
+        if peer_endpoint is not None:
+            peer_info = peer_endpoint.get_peer_info()
+            for active_unit in active_units.split(','):
```
Might want to allow for whitespace in case the admin sets the config to something like `"nfs/0, nfs/1"`.  `re.split(r'\w*,\w*', active_units)` should do it.

johnsca commented on reactive/nfs.py:
```diff
@@ -50,27 +54,96 @@ def update_config():
+        # that requirers know they need to unmount.
+        hookenv.log('No active units')
+        for relation in mount_interface.relations:
+            relation.to_publish_raw.update({
```
This is breaking the encapsulation of the interface layer and could cause problems on the off chance that the protocol is changed in the future.  Couldn't this data just be passed in to `mount_interface.configure()`?

johnsca reviewed (changes requested):
Mostly LGTM.  A few comments below, the only one of which I definitely think should be changed is the direct use of `to_publish_raw`; other comments are non-blockers / personal preferences, so feel free to ignore them if you disagree.

cjwatson committed:
Stop breaking the mount interface's encapsulation

cjwatson committed:
Simplify read_nfs_mounts prologue

cjwatson committed:
Simplify config handling in read_nfs_mounts

cjwatson committed:
Accept whitespace in active_units

cjwatson commented:
@johnsca Thanks; I've fixed all those issues.
